flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/inference/TypeInferenceOperandChecker.java

```diff
@@ -0,0 +1,164 @@
+
+	public TypeInferenceOperandChecker(
+			FunctionDefinition definition,
+			TypeInference typeInference) {
+		this.definition = definition;
+		this.typeInference = typeInference;
+		this.countRange = new ArgumentCountRange(typeInference.getInputTypeStrategy().getArgumentCount());
+	}
+
+	@Override
+	public boolean checkOperandTypes(SqlCallBinding callBinding, boolean throwOnFailure) {
+		final CallContext callContext = new CallBindingCallContext(definition, callBinding, null);
+		try {
+			return checkOperandTypesOrError(callBinding, callContext);
+		}
+		catch (ValidationException e) {
+			if (throwOnFailure) {
+				throw createInvalidCallException(callContext, e);
+			}
+			return false;
+		} catch (Throwable t) {
+			throw createUnexpectedException(callContext, t);
+		}
+	}
+
+	@Override
+	public SqlOperandCountRange getOperandCountRange() {
+		return countRange;
+	}
+
+	@Override
+	public String getAllowedSignatures(SqlOperator op, String opName) {
+		return TypeInferenceUtil.generateSignature(opName, definition, typeInference);
+	}
+
+	@Override
+	public Consistency getConsistency() {
+		return Consistency.NONE;
+	}
+
+	@Override
+	public boolean isOptional(int i) {
+		return false;
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private boolean checkOperandTypesOrError(SqlCallBinding callBinding, CallContext callContext) {
+		final AdaptedCallContext adaptedCallContext;
+		try {
+			adaptedCallContext = adaptArguments(
+				typeInference,
+				callContext,
+				null);
+		} catch (ValidationException e) {
+			throw createInvalidInputException(
+				typeInference,
+				callContext,
+				e);
+		}
+
+		insertImplicitCasts(callBinding, adaptedCallContext.getArgumentDataTypes());
+
+		return true;
+	}
+
+	private void insertImplicitCasts(SqlCallBinding callBinding, List<DataType> expectedDataTypes) {
+		final FlinkTypeFactory flinkTypeFactory = (FlinkTypeFactory) callBinding.getTypeFactory();
+		final List<SqlNode> operands = callBinding.operands();
+		for (int i = 0; i < operands.size(); i++) {
+			final LogicalType expectedType = expectedDataTypes.get(i).getLogicalType();
+			final RelDataType expectedRelDataType = flinkTypeFactory.createFieldTypeFromLogicalType(expectedType);
+
+			final SqlNode castedOperand = castTo(operands.get(i), expectedRelDataType);
+			callBinding.getCall().setOperand(i, castedOperand);
+			updateInferredType(callBinding.getValidator(), castedOperand, expectedRelDataType);
+		}
+	}
+
+	/**
+	 * Adopted from {@link org.apache.calcite.sql.validate.implicit.AbstractTypeCoercion}.
+	 */
+	private SqlNode castTo(SqlNode node, RelDataType type) {
+		return SqlStdOperatorTable.CAST.createCall(
+			SqlParserPos.ZERO,
+			node,
+			SqlTypeUtil.convertTypeToSpec(type).withNullable(type.isNullable()));
+	}
+
+	/**
+	 * Adopted from {@link org.apache.calcite.sql.validate.implicit.AbstractTypeCoercion}.
+	 */
+	@Deprecated
```

Comment:
The deprecated annotation was also adopted from Calcite. But I will remove it as it causes confusion.

Reply:
I think it will be better to remove it. Esp. as in Calcite it is actually not deprecated. There is only `  @SuppressWarnings("deprecation")` annotation.